designs/0004-closures-fun-types.md

```diff
@@ -0,0 +1,608 @@
+scope for `y` results in undefined behavior---the value of `y` is no
+longer available from the inner scope as we have returned from that
+scope.  The value of `f` persists outside of the local scope, but it
+contains a reference to a variable `y` that no longer exists.  This
+kind of dangling reference needs to be avoided.
+
+
+
+Explain the proposal as if it was already included in the project and you were teaching it to another Stan programmer in the manual. That generally means:
+
+- Introducing new named concepts.
+- Explaining the feature largely in terms of examples.
+- Explaining how Stan programmers should *think* about the feature, and how it should impact the way they use the relevant package. It should explain the impact as concretely as possible.
+- If applicable, provide sample error messages, deprecation warnings, or migration guidance.
+- If applicable, describe the differences between teaching this to existing Stan programmers and new Stan programmers.
+
+For implementation-oriented RFCs (e.g. for compiler internals), this section should focus on how compiler contributors should think about the change, and give examples of its concrete impact. For policy RFCs, this section should provide an example-driven introduction to the policy, and explain its impact in concrete terms.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Type system
+
+The underlying type system for Stan relies on a notion of runtime
+typing of all objects.  The runtime type does *not* include any
+constraints.  The constraints are used for bounds checking for read or
+constructed types and for transforms for parameters.
+
+Sized types are used for block declarations and unsized types are used
+for function arguments.  Local variables are currently sized, but in
+the future will be allowed to be unsized.
+
+
+#### Unsized runtime types
+
+The set of unsized runtime types is the least set of types such that
+
+* *unsized primitive type*: `int` and `real` are unsized types,
+* *unsized vector type*: `vector` and `row_vector` are unsized types,
+* *unsized matrix typen*: , `matrix` is an unsized type,
+* *unsized array type*: `T[]` is an unsized type if `T` is an unsized type, and
+* *unsized function type*: `T0(T1,...,TN)` is an unsized type if `T0`, `T1`, ..., `TN`
+are unsized types.
+
+
+
+#### Covariance and contravariance
+
+If we wanted to get fancy, we could use the polarity of type
+occurrences in function types to extend assignability.  For the base
+case, we have `int` as a subtype of `real` in the sense that we can
+assign an `int` expression to a `real` variable but not vice-versa.
+Covariant typing would extend this as follows.
+
+* `int` is a subtype `real`,
+* `T[]` is a subtype of `U[]` if `T` is a subtype of `U`, and
+* `T0(T1, ..., TN)` is a subtype of `U0(U1, ..., UN)` if
+  `T0` is a subtype of `U0`, `U1` is a subtype of `T1`, ..., `UN` is a
+  subtype of `TN`.
```

Comment:
I think this could be a separate issue on the stanc3 issue tracker - we could probably implement this subtyping in stanc3 right now without breaking anything or confusing anyone.